Add native block inserter onboarding tooltip

dcalhoun opened the pull request:
* **Gutenberg Mobile:** https://github.com/wordpress-mobile/gutenberg-mobile/pull/3529
* **WordPress iOS:** https://github.com/wordpress-mobile/WordPress-iOS/pull/16537
* **WordPress Android:** https://github.com/wordpress-mobile/WordPress-Android/pull/14683

## Description
Fixes #30624. Adds a tooltip for the block inserter as a part of a larger editor onboarding experience. 

A native implementation was added for the existing `Tooltip` component used on the web. This was done in an effort to (1) align the platforms and (2) avoid further duplicating tooltip logic. E.g. a duplicative tooltip component exists for the native [Cover media focal point picker](https://github.com/WordPress/gutenberg/blob/b955783c7b76a4518dbc47f867a4ff3cdf03f1d6/packages/components/src/focal-point-picker/tooltip/index.native.js#L24), and hopefully can one day be replaced by the `Tooltip` implementation in this PR.

The `Button` component [already renders the existing Tooltip component](https://github.com/WordPress/gutenberg/blob/b955783c7b76a4518dbc47f867a4ff3cdf03f1d6/packages/components/src/button/index.native.js#L188-L194). Because of this, it was not necessary to wrap the block inserter dropdown menu button in a new `Tooltip` element. Instead, this PR leverages `Button`'s `title` and `showTooltip ` props to display new "Tap to add content" tooltip required for the onboarding feature. One benefit of this approach is that we do not need to worry about forwarding a React ref through to `Button`. 

## How has this been tested?
Automated tests were not added in this PR, as the feature is an experiment and may not remain. However, I do plan to create a follow up PR to add tests for the `Tooltip` component if this PR merges. The work in this PR was manually tested via the following flow: 

1. Install the build from sibling WPiOS and WPAndroid PRs. 
1. Create/edit a post to launch the block editor. 
1. ℹ️ **Expected:** A "Tap to add content" tooltip is displayed attached to the block inserter button. 
1. Type on the keyboard. 
1. ℹ️ **Expected:** The tooltip remains visible and positioned correctly. 
1. Interact with any other portion of the UI aside from the keyboard or rotate the device. 
1. ℹ️ **Expected:** The tooltip is dismissed. 
1. Close the block editor. 
1. Relaunch the block editor. 
1. ℹ️ **Expected:** The tooltip is not displayed for subsequent editor launches. 

For Android, one can [clear out the WordPress app data](https://stackoverflow.com/a/8916939/378228) to re-test the flow. For iOS, the only option I am aware of is uninstalling and re-installing the app again. One thing to keep in mind is that the WPCom API places a limit on the number of times you may log in with a single account over a period of time. It will start a timeout if you log in too frequently. I'm uncertain of the number of attempts that triggers the timeout or how long the timeout lasts. 

## Screenshots 
| iOS | Android |
| --- | --- |
| ![ios-block-inserter-tooltip](https://user-images.githubusercontent.com/438664/118867255-7d3b8000-b8a8-11eb-9164-b0326f648dff.jpeg) | ![android-block-inserter-tooltip](https://user-images.githubusercontent.com/438664/118866638-cf2fd600-b8a7-11eb-9689-3fd39439f294.jpg) |

## Types of changes
New feature

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

dcalhoun committed:
[wip] Add block inserter tooltip

dcalhoun committed:
Fix linter warning

dcalhoun committed:
[wip] Initial native Tooltip component

Needs more accurate positioning, more styling, and better position prop
support.

dcalhoun committed:
[wip] Tooltip is seemingly aligning correctly

dcalhoun committed:
Reintroduce animation

dcalhoun committed:
Simplify Tooltip code

dcalhoun committed:
Allow Tooltip Slot to dismiss tooltips when touch occurs anywhere

The Tooltip Slot is currently successfully invoking the callback, but
the state is not triggering a re-render to hide the Tooltip.

dcalhoun committed:
Fix tooltip dismissal

Previously, _all_ tooltips registered a callback regardless if they were
visible. This meant that whichever tooltip was last was the only tooltip
registered. Often times this meant the currently visible tooltip would
remain left with a stale visibility as its callback had been cleared.

dcalhoun committed:
Only run tooltip animation when necessary

Previously, the animation ran anytime the visible prop changed, even if
it was the same value as the previous render.

dcalhoun committed:
Add note about performance considerations

dcalhoun committed:
Add documentation for `visible` prop

The `visible` prop was added as the hover/focus trigger used for web is
not relevant for touch devices. The ability to display the tooltip on
initial render is required for the current use case of an onboarding
experience.

dcalhoun committed:
Avoid tracking unused values

dcalhoun committed:
Clarify reference measuring timeout

Use a ref to avoid losing the timeout value. Rename the value to
increase clarity of intent.

dcalhoun committed:
Name ref explicitly

Include the phrase "ref" in the name to communicate the actual value.

dcalhoun committed:
Fix tooltip exit animation

dcalhoun committed:
Remove unused styles

dcalhoun committed:
Simplify render for hidden tooltip

dcalhoun committed:
Memoize context value to avoid unnecessary rerenders

Memoize object creation to avoid unnnecessary rerenders of the
Provider's children.

dcalhoun committed:
Adjust tooltip styles

dcalhoun committed:
Restrict editor onboarding tooltip visibility

Only display the editor onboarding tooltip for users who are first
launching the editor.

dcalhoun committed:
Refactor away redundant select invocations

Combing `select` calls to avoid multiple subscriptions that may(?)
degrade performance.

dcalhoun committed:
Revert "Refactor away redundant select invocations"

This reverts commit e4f091dc2c385b14dddfa88d8b6100c1465bb60d.

dcalhoun committed:
Relocate block inserter tooltip further down tree

To avoid the need to render a wrapping `View` or forward a `ref` down
the tree, the tooltip was moved further down the tree.

dcalhoun committed:
Leverage existing Tooltip within Button component

Rather than wrapping the block inserter button, we can leverage the
Tooltip component that is included inside of Button. In order to do so,
we must set `visible` whenever `showTooltip` is explicitly `true`.

dcalhoun committed:
Replace setTimeout with requestAnimationFrame

`requestAnimationFrame` feels more appropriate than `setTimeout` given
we are awaiting a render. This is required to await the render before
calculating the layout values to position the tooltip.

dcalhoun committed:
Reposition tooltip on keyboard visibility change

We must reposition the tooltip when the keyboard position changes to
avoid collisions.

dcalhoun committed:
Improve tooltip keyboard awareness

Reposition or hide the tooltip depending on the keyboard visibility.

dcalhoun committed:
Disable default tooltip visibility for native dropdown menu buttons

Enabling `showTooltip` within the dropdown menu was likely a copy from
the web implementation. Previously the mobile implementation of tooltip
was a no-op, rendering the `showTooltip` value moot. Now that tooltip
renders UI, we must disable `showTooltip` on mobile, otherwise every
dropdown menu button would display its tooltip on initial render.

dcalhoun committed:
Improve layout calculation resiliency

Now that we rely upon `requestAnimationFrame` to ensure the render
completes before calculation, there are times where the ref is no longer
present once the callback completes. This change ensures we check within
the `requestAnimationFrame` callback.

dcalhoun committed:
Improve visibility animation management

Add documentation and remove unnecessary hook dependency.

dcalhoun committed:
Manage tooltip position when keyboard frame changes

The keyboard frame (dimensions) can change when the current keyboard is
changed, the device rotates, dictation is toggled, etc. We must
reposition the tooltip when the frame updates.

dcalhoun committed:
Refactor tooltip styles

Hoist all styles underneath a `tooltip` selector. Prefer styles in Sass
files rather than JavaScript.

dcalhoun committed:
Avoid unnecessary touch handler when no tooltips displayed

Rather running a no-op anytime a touch is started, we can remove the
callback if no Tooltips are displayed.

dcalhoun committed:
Apply minor refactors

Done purely for stylistic preference.

dcalhoun commented on packages/components/src/button/index.native.js:
```diff
@@ -189,10 +192,11 @@ export function Button( props ) {
 			text={ label }
 			shortcut={ shortcut }
 			position={ tooltipPosition }
+			visible={ showTooltip === true }
```
An explicit check for `true` is used to determine whether the tooltip should be displayed on _initial render_. In the current implementation of the native Tooltip in this PR, initial render is the only mechanism for rendering a tooltip. However, future iterations could add interactions to display the tooltip (e.g. tap), which would be implemented within Tooltip itself.

dcalhoun commented on packages/components/src/button/index.native.js:
```diff
@@ -189,10 +192,11 @@ export function Button( props ) {
 		>
 			{ element }
 		</Tooltip>
 	);
 }
 
-export default withPreferredColorScheme( Button );
+export default Button;
```
`withPreferredColorScheme` was replaced with its sibling Hook implementation to avoid the need to forward refs through HoC's wrapping component.

dcalhoun replied:
The [previous comment](https://github.com/WordPress/gutenberg/pull/32001#discussion_r636392764) was once true and made passing a ref to `Button` easier, but that is no longer the case as this PR is now relying upon `Button`'s `title` and `showTooltip ` props to display the onboarding tooltip. So, in reality, I believe we could revert the changes to this file if we preferred. Although, I believe this project generally prefers Hooks over HoCs.

dcalhoun commented on packages/components/src/dropdown-menu/index.native.js:
```diff
@@ -124,7 +124,6 @@ function DropdownMenu( {
 						aria-haspopup="true"
 						aria-expanded={ isOpen }
 						label={ label }
-						showTooltip
```
`showTooltip` defaulting to `true` was likely carried over from the web implementation where the prop controls whether _hovering should display the button's tooltip_. Given that hovering is an interaction that is not possible on touch device currently, its presence makes less sense. 

With the previous native implementation of Tooltip being a no-op, this prop resulted in no noticeable impact. However, in the native Tooltip implemented in this PR, this prop would now cause the tooltip to be displayed on initial render for every button in the dropdown menu.

dcalhoun commented on packages/edit-post/src/components/layout/index.native.js:
```diff
@@ -117,53 +118,55 @@ class Layout extends Component {
-			>
-				<AutosaveMonitor disableIntervalChecks />
-				<View
+			<Tooltip.Slot>
```
`Tooltip.Slot` wraps the entire editor layout so that it may dismiss tooltips anytime touch input starts anywhere on the UI.

dcalhoun committed:
Revert refactor that caused error

An error occurred during creating or cleaning up the Hook.

dcalhoun commented:
@jhnstn @kyleaparker this work is ready for an initial design and code review. Installable builds may be found on the WPiOS and WPAndroid PRs referenced in the PR description. Would you be willing to take a look please? Thanks!

hypest commented:
I tested with the WPAndroid installable build and 🎉
✅ Expected: A "Tap to add content" tooltip is displayed attached to the block inserter button.
✅ Expected: The tooltip remains visible and positioned correctly.
✅ Expected: The tooltip is dismissed.
✅ Expected: The tooltip is not displayed for subsequent editor launches.

hypest commented on packages/block-editor/src/components/inserter/index.native.js:
```diff
@@ -31,16 +31,28 @@ import { store as blockEditorStore } from '../../store';
+} ) => (
 	<ToolbarButton
-		title={ __( 'Add block' ) }
+		title={
```
👋 @dcalhoun ! I see that the `title` prop is being reused(?) for hosting the tooltip content. Can you elaborate on this in the PR description, to add some info on what's going on here, how's the tooltip functionality implemented for native mobile for instance. A link to some resource that might already have the info would do it too. Does that make sense?

dcalhoun replied:
> 👋 @dcalhoun ! I see that the `title` prop is being reused(?) for hosting the tooltip content. Can you elaborate on this in the PR description, to add some info on what's going on here, how's the tooltip functionality implemented for native emobile for instance. A link to some resource that might already have the info would do it too. Does that make sense?

Definitely. I have updated the PR description with additional details as to _why_ these props are used, which I will copy here for convenience.

>The `Button` component [already renders the existing Tooltip component](https://github.com/WordPress/gutenberg/blob/b955783c7b76a4518dbc47f867a4ff3cdf03f1d6/packages/components/src/button/index.native.js#L188-L194). Because of this, it was not necessary to wrap the block inserter dropdown menu button in a new `Tooltip` element. Instead, this PR leverages `Button`'s `title` and `showTooltip ` props to display new "Tap to add content" tooltip required for the onboarding feature. One benefit of this approach is that we do not need to worry about forwarding a React ref through to `Button`. 

We could, of course, avoid this approach and merely wrap the existing `ToolbarButton` with `Tooltip` if that is preferred. We would just need to make sure the ref is forwarded to the `Button`, which currently is not set up. I reverted that support in https://github.com/WordPress/gutenberg/pull/32001/commits/ff738326f7edcec65686f4957f6376eb0293bc18#diff-4faa11aaadbb56229d147af80dfa880a68aa32b647c60c7690e710ef1d7db1d8.

Let me know what you think or if I failed to explain anything clearly. Thanks!

hypest replied:
Thanks for elaborating!

> We could, of course, avoid this approach and merely wrap the existing ToolbarButton with Tooltip if that is preferred.

FWIW, I'm cool with the current approach to reuse Button's title/showTooltip API. I was only interested in surfacing/documenting the implementation detail. 👍

hypest commented on packages/block-editor/src/components/inserter/index.native.js:
```diff
@@ -234,6 +247,7 @@ export class Inserter extends Component {
 		return (
 			<>
 				{ renderToggle( {
+					canViewEditorOnboarding,
```
Not related to the diff here per se but, am I the only one confused with which function `renderToggle()` is? :)

I mean, I see it coming both as a prop and also being a class member. If you are also confused, would it make sense to better differentiate the names as part of this PR? No strong feelings about it.

dcalhoun replied:
No, I too found myself retracing my steps a few times while deciphering the call stack of `renderToggle`. 😄 Happy to rename one of the instances for clarity.

hypest commented on packages/block-editor/src/components/inserter/index.native.js:
```diff
@@ -301,6 +315,7 @@ export default compose( [
 			getBlockOrder,
 			getBlockIndex,
 			getBlock,
+			getSettings,
```
Hmm, just to be sure, is this change (pulling up the `getSettings` destructuring) related to the tooltip feature, or just an opportunity to clean up the code in `getDefaultInsertionIndex()`? I see the change introduced in https://github.com/WordPress/gutenberg/commit/56dcae9110af78ae6e6ae19b55cf95f219e2bd16 but I feels not really related? 🤔

dcalhoun replied:
`getSettings` was hoisted higher up as it is now used in two places ([preexisting](https://github.com/WordPress/gutenberg/blob/1808871ceedcd07198eddf682686929ab5ed03b0/packages/block-editor/src/components/inserter/index.native.js#L339-L341), [new](https://github.com/WordPress/gutenberg/blob/1808871ceedcd07198eddf682686929ab5ed03b0/packages/block-editor/src/components/inserter/index.native.js#L381-L382)) due the need to access `canViewEditorOnboarding` in this file. My (untested) inclining was that "selecting" `getSettings` multiple times was possibly poorer performance. 

However, as you pointed out in https://github.com/WordPress/gutenberg/pull/32001#commitcomment-51125756, I overlooked actually using the new, destructured reference. 😄 I will update this file so both use the destructured reference, unless you'd prefer me to revert the destructuring. I do not have strong feelings on this topic. I'm sure the performance gain, if any, is marginal.

hypest replied:
Ah, I see. 👍 for going with the hoisted higher up destructured reference.

> I will update this file

While at it, wdyt about the other point in [my comment](https://github.com/WordPress/gutenberg/pull/32001#commitcomment-51125756) about renaming at the spot to something slightly more descriptive?

dcalhoun replied:
@hypest sure, happy to rename if it is helpful. However, there is [a preexisting `getEditorSettings`](https://github.com/WordPress/gutenberg/blob/6a09db457c0f3f937e214ccfd2a970dab81f24d0/packages/edit-post/src/components/header/header-toolbar/index.native.js#L119) that is used, albeit only once, alongside `getSettings` as well. Do you think the name `getEditorSettings` still makes sense in this file?

hypest replied:
> there is a preexisting getEditorSettings that is used

Good point. Looks like it's pulled from a different store than in our case here, maybe we can use the store name to make it more unique/specific: wdyt about naming it `getBlockEditorSettings()` here?

dcalhoun replied:
`getBlockEditorSettings` works for me! Thank you.

kyleaparker commented:
Great work @dcalhoun, this works as expected for me! 

Are we planning to show this to all users? I'm wondering if we should be excluding people who are already familiar with adding content to the editor, but I'm not sure how we can detect this.

dcalhoun commented:
> Great work @dcalhoun, this works as expected for me!

@kyleaparker thanks for testing!

> Are we planning to show this to all users? I'm wondering if we should be excluding people who are already familiar with adding content to the editor, but I'm not sure how we can detect this.

Sorry for the confusion. We are not showing this to all users. We plan to show this experience to a TBD percentage of users. However, the development/testing builds show this experience to _every_ install, purely to make development/testing easier. That is why you see the tooltip each time you re-install the test build. 

Additionally, last week we merged https://github.com/wordpress-mobile/gutenberg-mobile/pull/3437, which sets a "flag" whenever a user launches the Gutenberg block editor. So, if a user has launched the editor before, this flag would be set and they would not see the onboarding experience. If the flag is not set, they are considered a "new" user and would see the onboarding experience. 

Does that make sense? Happy to provide more details if it is helpful.

dcalhoun committed:
Rename class member to differentiate it from prop

The class member and prop were named identically, which may following
the flow a bit difficult.

dcalhoun committed:
Remove duplicative select

The `getSettings` value is already destructured earlier in the routine.

dcalhoun committed:
Rename getSettings to getBlockEditorSettings

Renamed to avoid ambiguity.

jhnstn commented on packages/block-editor/src/components/inserter/index.native.js:
```diff
@@ -321,11 +336,9 @@ export default compose( [
 			const {
 				__experimentalShouldInsertAtTheTop: shouldInsertAtTheTop,
-			} = getSettings();
+			} = getBlockEditorSettings();
```
👍  nice update!

jhnstn reviewed (commented):
I ran through the test steps on Android and iOS. 
I found everything worked as expected in Android.

I notice a few things on iOS:
- When opening a new post when the tooltip is active there is no focus on the title block. I expected the title block to be focused on when creating a new post.

- When rotating to landscape the tooltip disappears but when rotating back to portrait the tooltip re appears. Other events do fully disable the tooltip as expected. 

I wouldn't think the rotation disabling  is a blocker but might want to confirm the title focus change is ok.

I'm using an iPhone 11 iOS, v14.4

AmandaRiu reviewed (approved):
Hi @dcalhoun Tested WPiOS and WPAndroid - works great! Looked over the code and that looks really good as well, but I'm still learning and am not yet really familiar with the way SlotFills work and several other really interesting things in your code so I'd definitely feel better if there were a 2nd set of eyes on this one just to make sure I didn't miss anything. 👍

dcalhoun commented:
> * When opening a new post when the tooltip is active there is no focus on the title block. I expected the title block to be focused on when creating a new post.

@jhnstn this focus issue appears unrelated and to occur in the latest App Store release (17.3), but it only occurs for the _first launch after fresh installation._ I have been unable to reproduce the issue outside of a fresh install. Are you able to confirm you experience the same? If so, I would think we should address that separately from this work. WDYT? 

> * When rotating to landscape the tooltip disappears but when rotating back to portrait the tooltip re appears. Other events do fully disable the tooltip as expected.
> 
> I wouldn't think the rotation disabling is a blocker but might want to confirm the title focus change is ok.

Agreed that the rotation issue is minor and not worth blocking this work.

jhnstn commented:
@dcalhoun 
> Are you able to confirm you experience the same

Confirmed. I just installed the latest from the App store and I'm seeing the same focus flow.

jhnstn reviewed (approved):
Code looks good! 
My earlier testing observations are either unrelated or non blocking. 

:shipit:

dcalhoun committed:
Merge branch 'trunk' of github.com:WordPress/gutenberg into add/block-inserter-tooltip
